Review: реализация проекта stellar-burgers

kuliznikovjeka opened the pull request.

Commit:
feat: реализация stellar-burgers

- изменено название .env файла
- добавлен alias для селекторов в конфиг webpack
- настроен rtk и react-router
- реализован компонент обертка ProtectedRoute для защинных роутов
- настроена маршрутизация в компоненте app
- реализован запрос на ингридиенты и их отображение
- реализован компонент Informer для отображения ошибок или  информации  любой-другой
- реализована логика в компоненте Feed
- реализована логика входа, регистрации

Commit:
feat: реализация stellar-burgers

- изменен подход написания вложенных роутов в app.tsx
- добавлено css обнуление ссылок
- реализован процесс заказа бургера, сбор бургера
- реализованы содержимое модальных окон
- добавлены переходы по страницам в шапкке сайта, а также отображение активной ссылки
- добавлен запрос на заказы на странице профиля
- добавлены новые reducers в store
- реализованы слайсы: burgerConstructorSlice, makeOrderSlice, orderInfoSlice

Commit:
feat: реализация stellar-burgers

- убраны консоль-логи
- добавлены дынне для категории ингридиентов
- корректно настроено открытие модальных окон
- добавлен Link для логотипа
- добавлены прелоадеры
- реализована логика входа, восстановление пароля
- добавлены новые селекторы в userSlice

Commit:
feat: реализация stellar-burgers

- добавлен файл .env.example

kuliznikovjeka committed:
feat: реализация stellar-burgers

feat: реализация stellar-burgers

gennady-bars commented on src/components/burger-ingredient/burger-ingredient.tsx:
```diff
@@ -1,14 +1,18 @@
-import { FC, memo } from 'react';
+import { FC, memo, useCallback } from 'react';
+import { useDispatch } from 'react-redux';
```
* Хуки `useSelector` и `useDispatch` нужно импортировать из файла `store.ts` (они там типизированные уже), а не из библиотеки `react-redux`.  Нужно найти через поиск `react-redux` и заменить **все импорты**.

kuliznikovjeka replied:
Исправил)
Импорт из 'react-redux' был только в данном  файле

gennady-bars commented on src/components/app-header/app-header.tsx:
```diff
@@ -1,4 +1,9 @@
 import { AppHeaderUI } from '@ui';
+import { useSelector } from '../../services/store';
+import { userDataSelector } from '@slices';
 
```
**Можно лучше**


Если интересно, то можно создать **универсальную** функцию запроса `request` , которая будет содержать в себе уже **базовый урл**, проверку на **ok** и проверку на `success`, чтобы не дублировать их в каждом запросе.

```js

// 1 раз объявляем базовый урл
export const BASE_URL = "https://norma.nomoreparties.space/api/";

// создаем функцию проверки ответа на `ok`
const checkResponse = (res) => {
  if (res.ok) {
    return res.json();
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ошибка ${res.status}`);
};

// создаем функцию проверки на `success`
const checkSuccess = (res) => {
  if (res && res.success) {
    return res;
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ответ не success: ${res}`);
};

// создаем универсальную фукнцию запроса с проверкой ответа и `success`
// В вызов приходит `endpoint`(часть урла, которая идет после базового) и опции
const request = (endpoint, options) => {
  // а также в ней базовый урл сразу прописывается, чтобы не дублировать в каждом запросе
  return fetch(`${BASE_URL}${endpoint}`, options)
    .then(checkResponse)
    .then(checkSuccess);
};

```


И теперь не нужно будет дублировать все это в каждом запросе. Нужно просто заменить в запросах все `fetch` на `request`, удалив лишние проверки и базовый урл. Все остальное останется без изменений

Код станет намного чище. Будет мелкий аналог `axios`

Пример запроса ингредиентов

```js
const getIngredients =  () => request("ingredients");

```

В `get`-запросах даже не нужно указывать 2й параметр. Он получится `undefined`, что без проблем для `fetch` (который внутри).

gennady-bars commented on src/services/slices/orders/orders-slice.ts:
```diff
@@ -0,0 +1,51 @@
+  error: null
+};
+
+export const fetchOrders = createAsyncThunk('orders/fetchOrders', async () => {
+  const orders = await getOrdersApi();
+  return orders;
+});
```
**Можно лучше**

* Если внутри **безымянной** (стрелочной) функции вызывается одна функция с **точно такими же** аргументами, то эта безымянная функция не нужна. Это лишняя обертка вокруг самой функции, которую можно использовать просто сразу в коде


```js
export const fetchOrders = createAsyncThunk('orders/fetchOrders', getOrdersApi);


```

kuliznikovjeka replied:
Улучшил

gennady-bars commented on src/pages/login/login.tsx:
```diff
@@ -1,17 +1,40 @@
+  const error = useSelector(loginErrorSelector);
+  const isLoading = useSelector(userLoadingSelector);
+
   const [email, setEmail] = useState('');
   const [password, setPassword] = useState('');
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues={}) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

gennady-bars commented on src/pages/constructor-page/constructor-page.tsx:
```diff
@@ -1,20 +1,33 @@
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    dispatch(fetchIngredients());
```
`dispatch(fetchIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

kuliznikovjeka replied:
Исправил

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,196 @@
+  useEffect(() => {
+    dispatch(checkUserAuth());
+    dispatch(fetchIngredients());
+    dispatch(fetchFeeds());
```
`dispatch(fetchFeeds());` нужно удалить из `App`, так как получать заказы **ленты** нужно только при входе в **Ленту**, ведь пользователь может и не зайти на эту страницу

kuliznikovjeka replied:
Исправил

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,196 @@
+        </Route>
+      </Routes>
+
+      {backgroundLocation && (
```
* Отлично сделана логика отображения попапа и страницы  под ним

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:




* Отлично, что в `prepare` создаете `id`
* Хорошая логика попапов
* Отлично сделана логика отображения попапа и страницы  под ним
* Хорошо сделали `ProtectedRoute`


но есть некоторые недочеты:



* Хуки `useSelector` и `useDispatch` нужно импортировать из файла `store.ts` (они там типизированные уже), а не из библиотеки `react-redux`.  Нужно найти через поиск `react-redux` и заменить **все импорты**.
* `dispatch(fetchIngredients())` достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.
* `dispatch(fetchFeeds());` нужно удалить из `App`, так как получать заказы **ленты** нужно только при входе в **Ленту**, ведь пользователь может и не зайти на эту страницу 


**Можно лучше**

* Если внутри **безымянной** (стрелочной) функции вызывается одна функция с **точно такими же** аргументами, то эта безымянная функция не нужна. Это лишняя обертка вокруг самой функции, которую можно использовать просто сразу в коде

Commit:
feat: реализация stellar-burgers

- измен  путь импорта хука useDispatch на типизированный хук
- убраны лишние запросы
- убраны лишние обертки вокруг самих функций

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
